design/implicit-boot-mode.md

```diff
@@ -0,0 +1,197 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# implicit-boot-mode
+
+## Status
+
+provisional
+
+## Table of Contents
+
+<!--ts-->
+   * [Title](#title)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories [optional]](#user-stories-optional)
+            * [Story 1](#story-1)
+            * [Story 2](#story-2)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+The required boot mode for a host depends on several factors,
+including support in the hardware, whether IPv6 is being used, and
+whether virtual media is being used. We do not need to support all
+combinations of those features with all boot modes. Asking the user to
+give us the boot mode explicitly introduces an opportunity for them to
+specify an incorrect value. We have enough information about the host
+already to determine the correct boot mode to use. Therefore we should
+not need to extend the BareMetalHost API in order to support multiple
+boot modes.
+
+## Motivation
+
+When booting using virtual media, Ironic builds a custom ISO for each
+host. Today, it only includes the correct contents for one boot mode
+at a time. Some attempt is made to determine the right value by
+examining the hardware, but that is not always possible. When
+necessary, there is a way to configure a default.
+
+Several proposals have been floated to introduce a way for the user to
+explicitly set a boot mode. This proposal asserts that we can
+successfully manage the x86 hardware supported by metal3 today without
+it. Avoiding a new API field avoids several opportunities for
+misconfiguration.
```

Comment:
I suppose there wouldn't be a need to have explicit documentation, which says 'boot mode is configured automatically'. Since we would want to hide as much of Ironic stuff behind the API as possible. Yes?

Reply:
Explicit docs would really still be needed to explain what is done/supported in terms of the way Metal3 operates to set the operator expectation.

Reply:
Right, I think we can expand our driver documentation with a description of the boot mode that would be used, including any pre-configuration that might be needed for hosts that don't support probing or setting the boot mode via the BMC.

Reply:
That sounds good. I agree. Where is this driver documentation?

Reply:
> That sounds good. I agree. Where is this driver documentation?

There is some very minimal documentation in the API docs at https://github.com/metal3-io/baremetal-operator/blob/master/docs/api.md#spec-fields but I think we will want to add something new if we start having to explain to users that different drivers will have different behaviors.

Reply:
Yes, we will need that. Also, the doc above seems to be a description of the API object. Since the boot mode will _not_ be an extension to the API, documenting it in there would be tricky. How about we start another doc with non-API implementation details? Case in point, this boot-mode thing. And I'm sure there will be more as well, as we go along.